pkg/ipamd/ipamd.go

```diff
@@ -988,8 +997,24 @@ func (c *IPAMContext) nodeIPPoolReconcile(interval time.Duration) {
 
 func (c *IPAMContext) eniIPPoolReconcile(ipPool []string, attachedENI awsutils.ENIMetadata, eni string) {
 	seenIPs := make(map[string]bool)
+	attachedENIIPs := attachedENI.IPv4Addresses
+	isMetadataIPs := true
+	// Here we can't trust attachedENI since the IMDS metadata can be stale. We need to check with EC2 API.
+	// +1 is for the primary IP of the ENI that is not added to the ipPool and not available for pods to use.
+	if 1+len(ipPool) != len(attachedENIIPs) {
+		log.Warnf("Instance metadata does not match data store! ipPool: %v, metadata: %v", ipPool, attachedENIIPs)
+		log.Debugf("We need to check the ENI status by calling the EC2 control plane.")
+		// Call EC2 to verify IPs on this ENI
+		ec2Addresses, err := c.awsClient.GetIPv4sFromEC2(eni)
```

Comment:
How are we planning to testing this code path?

Reply:
Added a unit test step:
```
{"level":"debug","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd_test.go:669","msg":"IP Address Pool stats: total: 2, assigned: 0"}
{"level":"debug","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd_test.go:689","msg":"Reconciling ENI/IP pool info because time since last 81.451µs <= 0s"}
{"level":"debug","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd_test.go:689","msg":"Reconcile existing ENI eni-00000000 IP pool"}
{"level":"warn","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd.go:963","msg":"Instance metadata does not match data store! ipPool: [10.10.10.12 10.10.10.13], metadata: [{\n  Primary: true,\n  PrivateIpAddress: \"10.10.10.11\"\n}]"}
{"level":"debug","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd.go:963","msg":"We need to check the ENI status by calling the EC2 control plane."}
{"level":"error","ts":"2020-08-29T19:52:32.526-0700","caller":"ipamd/ipamd.go:963","msg":"Failed to fetch ENI IP addresses! Aborting reconcile of ENI eni-00000000"}
```

Reply:
I was asking if we can manually test the change once ? (Could be done by adding some sleep and triggering the code path)

Reply:
I think adding a temporary change where we just delete the last IP returned by IMDS should work to verify this code path.

Reply:
Debug build with the following change:
```
	log.Debugf("Reconcile existing ENI %s IP pool", attachedENI.ENIID)
	badIMDS := attachedENI
	addressCount := len(badIMDS.IPv4Addresses)
	if addressCount > 1 {
		badIMDS.IPv4Addresses = badIMDS.IPv4Addresses[:addressCount-1]
	}
	log.Debugf("XXXXX Deleted the last IP from attachedENI: %+v")
	// Reconcile IP pool
	c.eniIPPoolReconcile(eniIPPool, badIMDS)
```

```
{"level":"debug","ts":"2020-08-31T21:25:39.922Z","caller":"ipamd/ipamd.go:476","msg":"Reconciling ENI/IP pool info because time since last 1m0.127726217s <= 1m0s"}
{"level":"debug","ts":"2020-08-31T21:25:39.923Z","caller":"ipamd/ipamd.go:908","msg":"Total number of interfaces found: 2 "}
{"level":"debug","ts":"2020-08-31T21:25:39.923Z","caller":"awsutils/awsutils.go:568","msg":"Found ENI MAC address: 06:66:05:b5:c1:71"}
{"level":"debug","ts":"2020-08-31T21:25:39.924Z","caller":"awsutils/awsutils.go:582","msg":"Using device number 0 for primary eni: eni-01d689e8a83c6feba"}
{"level":"debug","ts":"2020-08-31T21:25:39.924Z","caller":"awsutils/awsutils.go:568","msg":"Found ENI: eni-01d689e8a83c6feba, MAC 06:66:05:b5:c1:71, device 0"}
{"level":"debug","ts":"2020-08-31T21:25:39.924Z","caller":"awsutils/awsutils.go:588","msg":"Found CIDR 192.168.128.0/19 for ENI 06:66:05:b5:c1:71"}
{"level":"debug","ts":"2020-08-31T21:25:39.924Z","caller":"awsutils/awsutils.go:588","msg":"Found IP addresses [192.168.131.160 192.168.149.75 192.168.134.205 192.168.128.237 192.168.150.15 192.168.152.224 192.168.134.35 192.168.129.132 192.168.156.197 192.168.131.24 192.168.134.155 192.168.147.91 192.168.159.31 192.168.133.50 192.168.152.119] on ENI 06:66:05:b5:c1:71"}
{"level":"debug","ts":"2020-08-31T21:25:39.924Z","caller":"awsutils/awsutils.go:568","msg":"Found ENI MAC address: 06:8f:fe:97:a2:9b"}
{"level":"debug","ts":"2020-08-31T21:25:39.925Z","caller":"awsutils/awsutils.go:568","msg":"Found ENI: eni-074c28aba8340f164, MAC 06:8f:fe:97:a2:9b, device 2"}
{"level":"debug","ts":"2020-08-31T21:25:39.926Z","caller":"awsutils/awsutils.go:588","msg":"Found CIDR 192.168.128.0/19 for ENI 06:8f:fe:97:a2:9b"}
{"level":"debug","ts":"2020-08-31T21:25:39.926Z","caller":"awsutils/awsutils.go:588","msg":"Found IP addresses [192.168.153.162 192.168.155.72 192.168.143.232 192.168.132.174 192.168.145.78 192.168.145.239 192.168.133.98 192.168.150.195 192.168.157.218 192.168.143.209 192.168.133.18 192.168.153.245 192.168.150.150 192.168.145.183 192.168.139.151] on ENI 06:8f:fe:97:a2:9b"}
{"level":"debug","ts":"2020-08-31T21:25:39.926Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-01d689e8a83c6feba IP pool"}
{"level":"debug","ts":"2020-08-31T21:25:39.926Z","caller":"ipamd/ipamd.go:476","msg":"XXXXX Deleted the last IP from attachedENI: %+v"}
{"level":"warn","ts":"2020-08-31T21:25:39.926Z","caller":"ipamd/ipamd.go:968","msg":"Instance metadata does not match data store! ipPool: [192.168.134.35 192.168.129.132 192.168.134.155 192.168.159.31 192.168.134.205 192.168.152.224 192.168.150.15 192.168.147.91 192.168.149.75 192.168.128.237 192.168.133.50 192.168.156.197 192.168.131.24 192.168.152.119], metadata: [{\n  Primary: true,\n  PrivateIpAddress: \"192.168.131.160\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.149.75\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.205\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.128.237\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.150.15\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.152.224\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.35\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.129.132\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.156.197\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.131.24\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.134.155\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.147.91\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.159.31\"\n} {\n  Primary: false,\n  PrivateIpAddress: \"192.168.133.50\"\n}]"}
{"level":"debug","ts":"2020-08-31T21:25:39.926Z","caller":"ipamd/ipamd.go:968","msg":"We need to check the ENI status by calling the EC2 control plane."}
{"level":"debug","ts":"2020-08-31T21:25:39.993Z","caller":"ipamd/ipamd.go:1026","msg":"Reconcile and skip primary IP 192.168.131.160 on ENI eni-01d689e8a83c6feba"}
{"level":"debug","ts":"2020-08-31T21:25:39.993Z","caller":"ipamd/ipamd.go:476","msg":"Reconcile existing ENI eni-074c28aba8340f164 IP pool"}
```